### ISSUE TITLE
testing CD script

### ISSUE THREAD

**Issue by edublancas** [ploomber/dummy#2]
Moving https://github.com/ploomber/jupysql/pull/104 here so we can iron out the details.

Once we're happy, we can add this to [pkgmt](https://github.com/ploomber/pkgmt) to apply it to our projects.

cc @yafimvo @idomic

**Comment by yafimvo** [comment]
@edublancas should I create a new PR and reference this issue? we need to update the script since this repo's master branch is named differentially (`main`).

**Comment by edublancas** [comment]
yes, feel free to open a PR to this repo with the changes. I copy pasted your script.

**Comment by edublancas** [comment]
works great, I just made a small change, so the command is `[release]`

@yafimvo: can you open a PR to https://github.com/ploomber/contributing/blob/main/maintainer.md so we document this? then, please open another to jupysql and sklearn-evaluation so start using this, I'll generate the tokens and add them to the repositories

**Comment by edublancas** [comment]
yeah, let's do ploomber-engine as well

**Comment by yafimvo** [comment]
@edublancas @idomic 
added a PR (couldn't add you as reviewers so mentioned it here as well)
https://github.com/ploomber/contributing/pull/20

**Comment by idomic** [comment]
@yafimvo we still need to add the part where it scans the changelog tags and releases if it's not [doc], so basically, it'll build automatically vs us needing to add the [release] tag

**Comment by edublancas** [comment]
there is another issue that's blocking us from using this. since we have branch protection (do not allow direct pushes to the main branch), the current script fails: https://github.com/ploomber/core/actions/runs/4058470460/jobs/6985484331

to fix this we can follow a similar approach as conda-forge: once a PR is merged, we can have a bot run `pkgmt version`, commit and open a PR, then, merging that PR can trigger `pkgmt release`.

Let's work on this branch protection issue first, and then tackle what Ido is mentioning.

**Comment by yafimvo** [comment]
I'm not sure I fully understand. I tried a few different approaches but none of them seems to work for me. Eventually, it fails when I run `pkgmt version` via the github actions if I have at least one protected branch in the repo.

I added a simple example where I run this command once a PR is merged. [This is the error it throws](https://github.com/ploomber/dummy/actions/runs/4065421232/jobs/7000094996).

It's weird since `pkgmt version` does work when I run it locally.

Solutions that I tried:
1. Create a specific branch for releases where we raise versions via a dedicated PR and merge it back to the master branch
2. Run it on the master branch

@edublancas @idomic any suggestions?

**Comment by yafimvo** [comment]
> I'm guessing the logs you shared are failing because the bot doesn't have permission to push directly to master due to the branch protections?

It seems like [allowing bot to push to protected branch](https://github.com/orgs/community/discussions/25305) is risky.

I found this discussion about [allowing github actions to bypass branch protection rules in certain specific circumstances](https://github.com/community/community/discussions/13836)

> also, one thing I realized is that we have to make a change to `pkgmt version` because it always pushes. so something like `pkgmt version --no-push`, and same for `pkgmt release`

I think this is the main issue here. I think if we create a PR for the version up + release without pushing it, and then approve and auto-merge it - it will work.

**Comment by edublancas** [comment]
> It seems like [allowing bot to push to protected branch](https://github.com/orgs/community/discussions/25305) is risky

the bot wont push. the bot will only open the PR

> I think this is the main issue here. I think if we create a PR for the version up + release without pushing it, and then approve and auto-merge it - it will work.

ok, let me make this change on pkgmt real quick

**Comment by edublancas** [comment]
> It seems like [allowing bot to push to protected branch](https://github.com/orgs/community/discussions/25305) is risky

the bot wont push. the bot will only open the PR

> I think this is the main issue here. I think if we create a PR for the version up + release without pushing it, and then approve and auto-merge it - it will work.

ok, I pushed pkgmt 0.1.6 with `pkgmt version --no-push`

**Comment by edublancas** [comment]
FYI: when the bot pushes to the branch used to open the PR, you need to push the tag as well, since it's not pushed automatically:

```
git push origin {tag}
```

**Comment by yafimvo** [comment]
The `pkgmt --no-push` works perfectly, however, there were still some challenges with the branch protection that prevented me from following this particular implementation approach (can't use the PR commit and needs approval).

So, I created the following workflow under `cd.yml`:
1. On a merge success to main branch with the new label `release` (instead of the [release] commit)
2. Create a new branch and run `pkgmt version --no-push`
3. Create a PR for release (with the new version)
4. @ploomber-bot approves this PR
5. Then merge this PR to main
6. Push the new tag
7. Release the version to PyPI

It works on my forked repo but fails on the original repo with a [permission error for github-actions[bot]](https://github.com/ploomber/dummy/actions/runs/4101702860/jobs/7073781517#step:7:72).

To make it easy to debug, I removed steps [5-7] so after labeling the PR with the `release` tag and merging it to the main branch - it should create a new PR with approval.

Since it's a permission issue, things I tried:
1. Make sure workflow permissions are set correctly - they are
![image](https://user-images.githubusercontent.com/103026689/216926278-cfb63a9c-1a2e-4d15-8a2e-3f2dd60c8c99.png)

2. Use an ACTIONS_PUSH_KEY with "god-mode" permissions instead GITHUB_TOKEN - [same error](https://github.com/ploomber/dummy/actions/runs/4101568231/jobs/7073497185#step:7:72)
3. add explicit permission to the `cd.yml` file - didn't work either

From googling it, maybe this is something with the package itself and we need to modify it manually but I don't have an access to this package page.

Found these:

https://github.com/orgs/community/discussions/26274#discussioncomment-3251137

https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions#installing-a-package-using-an-action


@edublancas what do you think?

**Comment by edublancas** [comment]
Seems like the issue is the auto-approving the PR.

I think approving it manually is okay; we can improve it later. If I understand correctly, the following pattern should not have any issues:

1. On a merge success to main branch with the new label release (instead of the [release] commit)
2. Create a new branch and run pkgmt version --no-push
3. Create a PR for release (with the new version)
4. ~@ploomber-bot approves this PR~ someone from our team approves the PR
5. Then merge this PR to main
6. Push the new tag
7. Release the version to PyPI

**Comment by yafimvo** [comment]
> * On a merge success to main branch with the new label release (instead of the [release] commit)
> * Create a new branch and run pkgmt version --no-push
> * Create a PR for release (with the new version)
> * ~@ploomber-bot approves this PR~ someone from our team approves the PR
> * Then merge this PR to main
> * Push the new tag
> * Release the version to PyPI

I added this flow to [workflows.yml](https://github.com/ploomber/dummy/blob/main/.github/workflows/workflows.yml) but since I was testing the flow, there is no open PR (it's already in the main branch).

The flow is:
1. User creates a PR for a new feature/bug fix/etc... - [example](https://github.com/ploomber/dummy/pull/25)
2. After it gets approved we can decide whether to release it by adding the `[release]` prefix to our commit message when merging with the main branch.
3. If we add the `[release]` prefix it will generate a release PR (checkout from the main) titled with the following tag version and labeled as `release`. This will also add specific team members as responsible for reviewing it (let's say @edublancas and @idomic) - [example](https://github.com/ploomber/dummy/pull/26)
4. After the approval - it will automatically push, release, and merge this PR with the main branch.


Few notes
1. The release PR might be outdated by the time the reviewer approves it. This may cause some conflicts on the merge. - I think we can automate the approval with @ploomber-bot (I think now with the new job dependencies it will work) and avoid the auto-merge. This way, we can create a [release] commit, and get notified when our release PR is auto-approved and passed the CI (or not).
2. Running `pkgmt release 0.1.13` releases the `0.1.14dev0` version. Should I run it the same as `pkgmt version --no-push` and then push the tag separately to PyPI?

@edublancas

**Comment by edublancas** [comment]
closing since I already fixed most of this with @yafimvo, I'll open a new issue with the newest updates
